src/tty-ask-password-agent/tty-ask-password-agent.c

```diff
@@ -614,8 +696,90 @@ static int parse_argv(int argc, char *argv[]) {
         return 1;
 }
 
+static unsigned int wfa_child(const struct console * con, const unsigned int id)
+{
+        setsid();
+        release_terminal();
+        *usemask |= 1 << id;   /* shared memory area */
+        current_dev = con[id].tty;
+        return id;
+}
+
+static unsigned int wait_for_answer(void)
+{
+        struct console *consoles;
+        struct sigaction sig = {
+                .sa_handler = chld_handler,
+                .sa_flags = SA_NOCLDSTOP | SA_RESTART,
+        };
+        struct sigaction oldsig;
+        sigset_t set, oldset;
+        unsigned int num = 0, id;
+        int status = 0, ret;
+        pid_t job;
+
+        consoles = collect_consoles(&num);
+        if (!consoles) {
+                log_error("Failed to query password: %m");
+                exit(EXIT_FAILURE);
+        }
+        if (num < 2)
+                return wfa_child(consoles, 0);
+
+        assert_se(sigemptyset(&set) == 0);
+        assert_se(sigaddset(&set, SIGHUP) == 0);
+        assert_se(sigaddset(&set, SIGCHLD) == 0);
+        assert_se(sigemptyset(&sig.sa_mask) == 0);
+        assert_se(sigprocmask(SIG_UNBLOCK, &set, &oldset) == 0);
+        assert_se(sigaction(SIGCHLD, &sig, &oldsig) == 0);
+        sig.sa_handler = SIG_DFL;
+        assert_se(sigaction(SIGHUP, &sig, NULL) == 0);
+
+        for (id = 0; id < num; id++) {
+                consoles[id].pid = fork();
+
+                if (consoles[id].pid < 0) {
+                        log_error("Failed to query password: %m");
+                        exit(EXIT_FAILURE);
+                }
+
+                if (consoles[id].pid == 0) {
+                        if (prctl(PR_SET_PDEATHSIG, SIGHUP) < 0)
+                                _exit(EXIT_FAILURE);
+                        zero(sig);
+                        assert_se(sigprocmask(SIG_UNBLOCK, &oldset, NULL) == 0);
+                        assert_se(sigaction(SIGCHLD, &oldsig, NULL) == 0);
+                        return wfa_child(consoles, id);
```

Comment:
The fork() is used to have child for each device (e.g. /dev/ttyS0 and /dev/tty1) which simply ask for the required password/passphrase.  If a password/passphrase is typed and entered (e.g. on /dev/ttyS0) then the code will kill the other asking child.  To get this synchronous a shared memory area is used, that is the index id is used to shift a bit for each running child.  Indeed the function wait_for_answer() returns in main for each child for its own whereas the parent process remains in wait_for_answer() to wait on the first successful password/passphrase submission

A similar code I had written some times ago for sulogin of  upstream util-linux. It does work flawless as well I've tested this change for systemd several times successfully.